app/player/cache.go

```diff
@@ -0,0 +1,199 @@
+	var size int64
+	fi, _ := os.Stat(s.getPath(hash))
+	if fi != nil {
+		size = fi.Size()
+	}
+
+	if err := os.Remove(s.getPath(hash)); err != nil {
+		CacheLogger.Log().Errorf("failed to evict chunk file %v: %v", hash, err)
+		return
+	}
+	metrics.PlayerCacheSize.Sub(float64(size))
+	CacheLogger.Log().Infof("chunk file %v evicted", hash)
+}
+
+func (s fsStorage) getPath(hash interface{}) string {
+	return path.Join(s.path, hash.(string))
+}
+
+func (s fsStorage) open(hash interface{}) (*os.File, error) {
+	f, err := os.Open(s.getPath(hash))
+	if err != nil {
+		return nil, err
+	}
+	return f, nil
+}
+
+// Has returns true if cache contains the requested chink.
+// It is not guaranteed that actual file exists on the filesystem.
+func (c *fsCache) Has(hash string) bool {
+	_, ok := c.rCache.Get(hash)
+	return ok
+}
+
+// Get returns ReadableChunk if it can be retrieved from the cache by the requested hash
+// and a boolean representing whether chunk was found or not.
+func (c *fsCache) Get(hash string) (ReadableChunk, bool) {
+	if value, ok := c.rCache.Get(hash); ok {
+		f, err := c.storage.open(value)
+		if err != nil {
+			metrics.PlayerCacheErrorCount.Inc()
+			CacheLogger.Log().Errorf("chunk %v found in cache but couldn't open the file: %v", hash, err)
+			c.rCache.Del(value)
+			return nil, false
+		}
+		metrics.PlayerCacheHitCount.Inc()
+		cb, err := initCachedChunk(f)
+		f.Close()
+		if err != nil {
+			CacheLogger.Log().Errorf("chunk %v found in cache but couldn't read the file: %v", hash, err)
+			return nil, false
+		}
+		return cb, true
+	}
+	metrics.PlayerCacheMissCount.Inc()
+	CacheLogger.Log().Debugf("cache miss for chunk %v", hash)
+	return nil, false
+}
+
+// Set takes chunk body and saves reference to it into the cache table
+func (c *fsCache) Set(hash string, body []byte) (ReadableChunk, error) {
+	cacheCost := len(body)
+
+	CacheLogger.Log().Debugf("attempting to cache chunk %v", hash)
+	chunkPath := c.storage.getPath(hash)
+	f, err := os.OpenFile(chunkPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600)
+	if os.IsExist(err) {
+		metrics.PlayerCacheErrorCount.Inc()
+		CacheLogger.Log().Debugf("chunk %v already exists on the local filesystem, not overwriting", hash)
+	} else {
+		numWritten, err := f.Write(body)
+		f.Close()
+		if err != nil {
+			metrics.PlayerCacheErrorCount.Inc()
+			CacheLogger.Log().Errorf("error saving cache file %v: %v", chunkPath, err)
+			return nil, err
+		}
+		CacheLogger.Log().Debugf("written %v bytes for chunk %v", numWritten, hash)
+	}
+	c.rCache.Set(hash, hash, int64(cacheCost))
+	// metrics.PlayerCacheSize.Set(float64(c.rCache.Metrics.CostAdded() - c.rCache.Metrics.CostEvicted()))
+
+	metrics.PlayerCacheSize.Add(float64(cacheCost))
+	CacheLogger.Log().Debugf("chunk %v successfully cached", hash)
+	return &cachedChunk{reflectedChunk{body}}, nil
+}
+
+// Remove deletes both cache record and chunk file from the filesystem.
+func (c *fsCache) Remove(hash string) {
```

Comment:
This should return an error if it fails to remove. Also the things inside should return errors to check.